hy/compiler.py

```diff
@@ -598,6 +621,166 @@ def compile_raise_expression(self, expr, root, exc, cause):
             expr, type=ret.expr, exc=exc,
             inst=None, tback=None, cause=cause)
 
+    notsym = lambda *dissallowed: some(
+        lambda x: isinstance(x, Symbol) and str(x) not in dissallowed
+    )
+    keepsym = lambda wanted: some(lambda x: isinstance(x, Symbol) and x == wanted)
+    _match_value = SYM | LITERAL
+    _pattern = forward_decl()
+    _assignment = maybe(sym(":as") + SYM)
+    _pattern.define(
+        _match_value + _assignment
+        | brackets(many(_pattern | unpack("iterable"))) + _assignment
+        | pexpr(keepsym("|"), many(_pattern)) + _assignment
+        | pexpr(keepsym(","), many(_pattern)) + _assignment
+        | curly_brackets(many(LITERAL + _pattern), maybe(pvalue("unpack-mapping", SYM))) + _assignment
+        | pexpr(
+            notsym("|", ",", "unpack-mapping", "unpack-iterable"),
+            many(_pattern),
+            many(KEYWORD + _pattern),
+        ) + _assignment
+    )
+    match_clause = _pattern + maybe(sym(":if") + FORM) + _assignment
+
+    @special(
+        (PY3_10, "match"), [FORM, oneplus(pexpr(sym("case"), match_clause, many(FORM)))]
```

Comment:
Agreed with @Kodiologist we should drop the "case"; and a mild preference for the unbracketed version of the syntax. Also this should be `many` instead of `oneplus` to match other parts of Hy that allow vacuous constructs.

Reply:
valid, done.